File: packages/py/examples/support_agent/in_memory_demo.py
"""Runs the whole support flow in one process over an in-memory bus: the agent
drafts and acts; the approver auto-approves the sensitive steps; the runner
resumes. Deterministic — no network, no keys."""

import asyncio

from airlock import (
    Agent,
    AgentDependencies,
    AgentRunner,
    Approver,
    FakeLlmProvider,
    InMemoryAuditSink,
    InMemoryEventBus,
    InMemoryRunStore,
    MessageRole,
    RiskBasedGatePolicy,
    SystemClock,
    UuidIdGenerator,
    auto_approve_decision_source,
)
from airlock.domain.run import RunState
from examples.support_agent.wiring import (
    SUPPORT_REQUEST,
    SUPPORT_SCRIPT,
    SUPPORT_SYSTEM_PROMPT,
    create_support_tools,
)


def _log(message: str) -> None:
    print(message)


def _final_answer(state: RunState | None) -> str:
    if state is None or not state.messages:
        return "(no answer)"
    last = state.messages[-1]
    return last.content if last.role == MessageRole.ASSISTANT else "(no answer)"


async def main() -> None:
    _log("=== Airlock support-agent demo (in-memory) ===")
    bus = InMemoryEventBus()
    store = InMemoryRunStore()
    audit = InMemoryAuditSink()
    agent = Agent(
        AgentDependencies(
            provider=FakeLlmProvider(SUPPORT_SCRIPT),
            tools=create_support_tools(_log),
            events=bus,
            store=store,
            audit=audit,
            clock=SystemClock(),
            ids=UuidIdGenerator(),
            gate_policy=RiskBasedGatePolicy(),
            system_prompt=SUPPORT_SYSTEM_PROMPT,
        )
    )
    await AgentRunner(agent, bus).start()
    await Approver(bus, bus, auto_approve_decision_source("demo-operator", _log)).start()

    _log(f"\n[USER] {SUPPORT_REQUEST}\n")
    started = await agent.run(SUPPORT_REQUEST)
    final = await store.load(started.run_id)

    _log(f"\n[AGENT] {_final_answer(final)}")
    status = final.status if final is not None else "unknown"
    _log(f"\n[RESULT] status={status} audit_events={len(audit.all())}")


if __name__ == "__main__":
    asyncio.run(main())

File: packages/ts/examples/support-agent/in-memory-demo.ts
import {
  Agent,
  AgentRunner,
  Approver,
  autoApproveDecisionSource,
  FakeLlmProvider,
  InMemoryAuditSink,
  InMemoryEventBus,
  InMemoryRunStore,
  MessageRole,
  RiskBasedGatePolicy,
  type RunState,
  SystemClock,
  UuidIdGenerator,
} from "../../src/index";
import {
  createSupportTools,
  SUPPORT_REQUEST,
  SUPPORT_SCRIPT,
  SUPPORT_SYSTEM_PROMPT,
} from "./wiring";

function log(message: string): void {
  process.stdout.write(`${message}\n`);
}

function finalAnswer(state: RunState | null): string {
  const last = state?.messages.at(-1);
  return last && last.role === MessageRole.Assistant ? last.content : "(no answer)";
}

/**
 * Runs the whole support flow in one process over an in-memory bus: the agent
 * drafts and acts; the approver auto-approves the sensitive steps; the runner
 * resumes. Deterministic — no network, no keys.
 */
async function main(): Promise<void> {
  log("=== Airlock support-agent demo (in-memory) ===");
  const bus = new InMemoryEventBus();
  const store = new InMemoryRunStore();
  const audit = new InMemoryAuditSink();
  const agent = new Agent({
    provider: new FakeLlmProvider(SUPPORT_SCRIPT),
    tools: createSupportTools(log),
    events: bus,
    store,
    audit,
    clock: new SystemClock(),
    ids: new UuidIdGenerator(),
    gatePolicy: new RiskBasedGatePolicy(),
    systemPrompt: SUPPORT_SYSTEM_PROMPT,
  });
  await new AgentRunner({ agent, subscriber: bus }).start();
  await new Approver({
    publisher: bus,
    subscriber: bus,
    decide: autoApproveDecisionSource("demo-operator", log),
  }).start();

  log(`\n[USER] ${SUPPORT_REQUEST}\n`);
  const started = await agent.run(SUPPORT_REQUEST);
  const final = await store.load(started.runId);

  log(`\n[AGENT] ${finalAnswer(final)}`);
  log(`\n[RESULT] status=${final?.status ?? "unknown"} audit_events=${String(audit.all().length)}`);
}

main().catch((error: unknown) => {
  process.stderr.write(`${String(error)}\n`);
  process.exitCode = 1;
});
